osbrain/tests/test_agent_sync_publications.py

```diff
@@ -73,6 +81,175 @@ def on_error(agent):
+        client.subscribe('sub', handler={'normal': append_received})
+    else:
+        client.subscribe('sub', handler={'negate': receive_negate,
+                                         'normal': append_received})
+
+    server.each(0.1, 'publish')
+
+    assert wait_agent_condition(client, match_tail, 8, [-1, 2] * 4)
+
+
+@pytest.mark.parametrize('socket_type', ['PUB', 'SYNC_PUB'])
+def test_resubscribe(nsproxy, socket_type):
+    '''
+    After subscribing to the same topic with a different handler after a
+    subscription to that topic was already made, we should override the
+    original handler.
+
+    Note that we will only override explicitly given topics (previous
+    subscriptions of other topics will remain untouched).
+    '''
+    server = run_agent('server', base=ServerNumbers)
+    client = run_agent('client')
+
+    addr = server.bind(socket_type, alias='pub', handler=lambda: None)
+    client.set_attr(received=[])
+    client.connect(addr, alias='sub', handler=append_received)
+
+    # Establish a connection
+    server.each(0.1, 'send', 'pub', 'connecting...', alias='tmp')
+    assert wait_agent_attr(client, data='connecting...')
+    server.stop_timer('tmp')
+
+    # Subscribe to two topics
+    client.unsubscribe('sub', '')
+    client.subscribe('sub', handler={'negate': receive_negate,
+                                     'normal': append_received})
+
+    server.each(0.1, 'publish')
+
+    assert wait_agent_condition(client, match_tail, 8, [-1, 2] * 4)
+
+    # Resubscribe, so as to store the 'normal' (2) messages negated
+    client.subscribe('sub', handler={'normal': receive_negate})
+
+    assert wait_agent_condition(client, match_tail, 8, [-1, -2] * 4)
+
+
+@pytest.mark.parametrize('socket_type', ['PUB', 'SYNC_PUB'])
+def test_unsubscribe_and_subscribe_again(nsproxy, socket_type):
+    '''
+    Test resubscribing to topics after unsubscribing works fine.
+    '''
+    server = run_agent('server', base=ServerNumbers)
+    client = run_agent('client')
+
+    addr = server.bind(socket_type, alias='pub', handler=lambda: None)
+    client.set_attr(received=[])
+    client.connect(addr, alias='sub', handler=append_received)
+
+    # Establish a connection
+    server.each(0.1, 'send', 'pub', 'connecting...', alias='tmp')
+    assert wait_agent_attr(client, data='connecting...')
+    server.stop_timer('tmp')
+
+    # Subscribe to two topics
+    client.unsubscribe('sub', '')
+    client.subscribe('sub', handler={'negate': receive_negate,
+                                     'normal': append_received})
+
+    # Make server to start publishing messages
+    server.each(0.1, 'publish')
+
+    assert wait_agent_condition(client, match_tail, 8, [-1, 2] * 4)
+
+    # Unsubscribe and reset received messages
+    client.unsubscribe('sub', ['negate', 'normal'])
+    client.set_attr(received=[])
+
+    # Nothing should be received (length == 0)
+    assert not wait_agent_attr(client, length=1, timeout=0.5)
```

Comment:
Why don't you unsubscribe from one of the topics only so that you can do:

```python
assert wait_agent_attr(client, match_tail, [-1] * 8)
```

instead?

That means, positive assertion, with no need for a timeout (less risk). You can decrease the publisher period to `0.01` to make things faster.

Also, this way you do not need to reset the `received` attribute. :wink:

Reply:
At first I did what you proposed, but then the code should be ``client.unsubscribe('sub', 'normal')`` 😅 . So I changed it for the other way around, just to avoid any subliminal messages hahaha.